### DISCUSSION TITLE
Uses step in Composite Action

### DISCUSSION THREAD

**Discussion by vesubramanian** [community/community#25778] [Actions]
<p>I am trying to build a Composite action. In that I am trying to use another action with “uses” statement. But I am getting an error when I try to call this composite action from a workflow.- “Unexpected value ‘uses’”. Can’t we call another action from within a composite action? Couldn’t find it in documentation also.</p>

**Comment by mickeygousset** [comment, marked as answer]
<p>Composite Actions can only be created using run steps. They are only designed to chain together multiple “run” steps. <a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action" class="inline-onebox">Creating a composite run steps action - GitHub Docs</a></p>

**Comment by laughedelic** [comment]
<p>This has been discussed multiple times on this forum, here is an incomplete list of relevant topics:</p>
<ul>
<li><a href="https://github.community/t/templates-for-github-actions/18045">Templates for GitHub Actions? </a></li>
<li><a href="https://github.community/t/reusable-component-composite-action-python-version/137833">Reusable component composite action python version</a></li>
<li><a href="https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851">reusing/sharing/inheriting steps between jobs declarations </a></li>
<li><a href="https://github.community/t/support-for-yaml-anchors/16128">Support for YAML anchors</a></li>
<li><a href="https://github.community/t/reuse-github-actions-inside-other-actions/122312">Reuse Github Actions inside other Actions? </a></li>
<li><a href="https://github.community/t/is-it-possible-to-reuse-workflow-yaml-to-setup-similar-workflows/17142">Is it possible to reuse workflow yaml to setup similar workflows? </a></li>
<li><a href="https://github.community/t/declaratively-define-github-actions/119684">Declaratively define GitHub actions </a></li>
<li><a href="https://github.community/t/common-steps-for-all-workflows/127272">Common steps for all workflows? </a></li>
<li><a href="https://github.community/t/executing-external-github-action-from-shell-or-js-scripts/17003">Executing external GitHub action from shell or JS scripts</a></li>
<li><a href="https://github.community/t/allow-an-action-which-groups-other-actions/16564">Allow an action which groups other actions </a></li>
</ul>

**Comment by vesubramanian** [comment]
<p>Thank you. Unfortunately, the documentation doesn’t state what is supported &amp; not supported. This is the reason for the confusion.</p>

**Comment by aeneasr** [comment]
<p>Are there any plans to support <code>use</code> in composite actions? Without those, they’re just another way of writing a bash script :frowning:</p>

**Comment by airtower-luna** [comment]
<p>There is some ongoing work:</p><aside class="onebox githubpullrequest" data-onebox-src="https://github.com/actions/runner/pull/1144">
  <header class="source">

      <a href="https://github.com/actions/runner/pull/1144" target="_blank" rel="noopener">github.com/actions/runner</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/runner/pull/1144" target="_blank" rel="noopener">Composite Actions Support ADR</a>
    </h4>

    <div class="branches">
      <code>actions:main</code> ← <code>actions:thboop/CompositeAdr</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-06-10" data-time="21:41:30" data-timezone="UTC">09:41PM - 10 Jun 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/thboop" target="_blank" rel="noopener">
          <img alt="thboop" src="https://user-images.githubusercontent.com/5814919/181093242-b0fa58b5-a3db-464f-ad4d-cd580cf88dc1.jpeg" class="onebox-avatar-inline" width="20" height="20">
          thboop
        </a>
      </div>

      <div class="lines" title="3 commits changed 1 files with 94 additions and 0 deletions">
        <a href="https://github.com/actions/runner/pull/1144/files" target="_blank" rel="noopener">
          <span class="added">+94</span>
          <span class="removed">-0</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">[Rendered](https://github.com/actions/runner/blob/4d7988e0ad30584f7c61d8b669a6ba<span class="show-more-container"><a href="https://github.com/actions/runner/pull/1144" target="_blank" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">bf04d0e486/docs/adrs/0000-composite-actions.md)
We welcome community feedback, please feel free to comment on this ADR! 

This ADR outlines the next steps we are taking to change composite run steps into composite actions, and specifically mentions how certain scenarios will work.</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>I haven’t seen anything about when we can expect this to become available, though.</p>

**Comment by JamesArthurHolland** [comment]
<p>Shocking amounts of code duplication because this isn’t a feature</p>

**Comment by ramsey** [comment]
<p>According to the Metadata Syntax documentation, under the section for "<code>runs</code> for composite actions," composite actions allow the use of <code>uses</code>. See here: <a href="https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsstepsuses" class="inline-onebox" rel="noopener nofollow ugc">Metadata syntax for GitHub Actions - GitHub Docs</a></p>
<p>Is this a recent change, or is the documentation incorrect?</p>
<p>If the documentation is correct and this works now, I’m still not able to get it working. I’m seeing <code>Unexpected value 'uses'</code> errors. (for example, here: <a href="https://github.com/ramsey/composer-install/actions/runs/1613059026" class="inline-onebox" rel="noopener nofollow ugc">run actions on all pushes · ramsey/composer-install@793876b · GitHub</a>)</p>
<hr>
<p>The documentation shows examples like this:</p>
<pre><code class="lang-yaml">runs:
  using: "composite"
  steps:
    # Reference a specific commit
    - uses: actions/checkout@a81bbbf8298c0fa03ea29cdc473d45769f953675
    # Reference the major version of a release
    - uses: actions/checkout@v2
    # Reference a specific version
    - uses: actions/checkout@v2.2.0
    # Reference a branch
    - uses: actions/checkout@main
    # References a subdirectory in a public GitHub repository at a specific branch, ref, or SHA
    - uses: actions/aws/ec2@main
    # References a local action
    - uses: ./.github/actions/my-action
    # References a docker public registry action
    - uses: docker://gcr.io/cloud-builders/gradle
    # Reference a docker image published on docker hub
    - uses: docker://alpine:3.8
</code></pre>
<p>and this:</p>
<pre><code class="lang-yaml">runs:
  using: "composite"
  steps:
    - name: My first step
      uses: actions/hello_world@main
      with:
        first_name: Mona
        middle_name: The
        last_name: Octocat  
</code></pre>

**Comment by ramsey** [comment]
<p>After some trial-and-error, I got <code>uses</code> working with composite actions. They were failing because I was specifying <code>working-directory</code>. Once I removed that, they work.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="ramsey" data-post="8" data-topic="181289">
<div class="title">
<div class="quote-controls"></div>
 ramsey:</div>
<blockquote>
<p>Is this a recent change, or is the documentation incorrect?</p>
</blockquote>
</aside>
<p>Sort of recent, as in August, about one month after the last post in this thread before yours. :wink:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">

      <a href="https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/" target="_blank" rel="noopener">The GitHub Blog</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/" target="_blank" rel="noopener">GitHub Actions: Reduce duplication with action composition | GitHub Changelog</a></h3>

  <p>GitHub Actions: Reduce duplication with action composition</p>

  <p>
    <span class="label1">Est. reading time: 1 minute</span>
  </p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
